### PR TITLE
RISCV: Backport some upstream changes

### PR DESCRIPTION
**Pull request by Ast-x64**
Based on https://github.com/dlang/phobos/pull/8477,
https://github.com/dlang/phobos/pull/8478, and
https://github.com/dlang/phobos/pull/8479.
Also adjust the order of some version conditions in std/math/hardware.d
to allow compilation using a RISCV64 port of LDC.

### CODE REVIEW

File: std/math/hardware.d [hunk 1]
```diff
@@ -905,7 +905,18 @@ private:
     // Read from the control register
     package(std.math) static ControlState getControlState() @trusted pure
     {
-        version (LDC)
+        version (RISCV_Any)
+        {
+            mixin(`
```
Comment (JohanEngelen) [inline std/math/hardware.d:910]: why is a mixin used here?

Comment (Ast-x64) [reply std/math/hardware.d:910]: Thanks for your review. I'm not so sure why the previous implementation uses mixin. All I've done is moving this `RISCV_Any` condition above the `LDC` condition. It was first committed in https://github.com/dlang/phobos/commit/3d84369f87be8a341772f5356569500fdf25aba1.

Comment (kinke) [reply std/math/hardware.d:910]: IIRC, that was a problem of the dscanner linter not supporting GDC-style inline asm...

Comment (JohanEngelen) [reply std/math/hardware.d:910]: ah ok. Well it's OK. Refactoring is usually better done separately :)

**Comment by kinke** [comment]
I've added the missing LDC-specific asm blocks; the other stuff had already been added (I haven't checked the unittests).

**Pull request closed without merging**